dynn/boosted_wrapper.py

```diff
@@ -0,0 +1,84 @@
+import torch
+import torch.nn as nn
+from grit.head.san_graph import SANGraphHead
+from torch_geometric.graphgym.models import GNNGraphHead
+from gradient_rescale import GradientRescaleFunction
+import numpy as np
+
+class BoostedWrapper(nn.Module):
+
+    def __init__(self, grit_transformer, head_dim_in, head_dim_out, num_classes = 10, ensemble_reweight = [0.5]):
+        super().__init__()
+        self.grit_transformer = grit_transformer
+        self.head_dim_in = head_dim_in
+        self.head_dim_out = head_dim_out
+        self.ensemble_reweight = ensemble_reweight
+        self.num_classes = num_classes
+
+    def set_intermediate_heads(self, intermediate_head_positions):
+        self.intermediate_head_positions = intermediate_head_positions
+        self.num_of_gates = len(intermediate_head_positions)
+        self.num_layers = len(intermediate_head_positions) + 1
+        head = self.grit_transformer.post_mp
+        assert self._is_supported_graph_head(type(head).__name__), "Unsupported head type, make sure you add support for it"
+        self.intermediate_heads = nn.ModuleList([
+            SANGraphHead(dim_in = head.dim_in,
+                         dim_out=head.dim_out,
+                         L=head.L) if type(head).__name__ == 'SANGraphHead' else GNNGraphHead(self.head_dim_in, self.head_dim_out)
+            for _ in range(len(self.intermediate_head_positions))])
+        assert len(self.ensemble_reweight) in [1, 2, self.num_layers]
+        if len(self.ensemble_reweight) == 1:
+            self.ensemble_reweight = self.ensemble_reweight * self.num_layers
+        elif len(self.ensemble_reweight) == 2:
+            self.ensemble_reweight = list(np.linspace(self.ensemble_reweight[0], self.ensemble_reweight[1], self.num_layers))
+
+
+    def boosted_forward(self, batch):
+        res = []
+        for name, module in self.grit_transformer.named_children():
+            if name == 'layers':
+                for layer_idx in range(len(module)):
+                    intermediate_layer = module[layer_idx]
+                    batch = intermediate_layer(batch)
+                    # TODO fix the out of bound issue
+                    last_idx = len(batch) - 1
```

Comment:
I can't remember why this is done but they also do it in boosted. They index the batch at the last element. See here [x[-1] here](https://github.com/SHI-Labs/Boosted-Dynamic-Networks/blob/main/models/msdnet_ge.py#L376).

When using graph gym since this is not a regular batch but rather a databatch, you can't index the last element using [-1] so i simply find the last index like that.